**Importing necessary libraries**

In [18]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

**Import the CSV data file**

In [19]:
df = pd.read_csv("data/StudentsPerformance.csv")

**Show Top 10 rows**

In [20]:
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78  
6               completed          88             95             92  
7                    none          40             43             39  
8               completed          64             64             67  
9                    none          38             60             50

**Shape of the dataframe**

In [21]:
df.shape

(1000, 8)

**DATA DESCRIPTION**

1. Gender	                    : Student's gender (Male/Female)
2. Race/Ethnicity	            : Race information (Grouped as Group A, B, C, D)
3. Parental level of education	: Parents educational information (Divided into bachelor's degree,
                                  some college, master's degree, associate's degree, high school)
4. Lunch	                    : Had lunch before test (Standard or Free/Reduced)
5. Test preparation course	    : Preparation course completed or not
6. Math score	                : Score in maths
7. Reading score	            : Score in reading
8. Writing score                : Score in writing

**DATA CHECKS TO PERFORM**

1. Check missing values
2. Check duplicates
3. Check data type
4. Check the number of unique values of each column
5. Check statistics of data set
6. Check various categories present in different categorical column

**1. Check missing values** 

In [22]:
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

No missing values in the dataset

**2. Check duplicates**

In [23]:
df.duplicated().sum()

0

No duplicate values in the dataset

**3. Check data type of columns**

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


**4. Check number of unique values in each column**

In [25]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

**5. Check statistics of data**

In [26]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

**Statistical insight**
1) In case of numerical data, the mean values for the three subjects are very close to each other - between 66 and 68.05
2) Similarly standard deviations are also close - between 14.6 and 15.19
3) While there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17

**EXPLORING THE DATA**

In [27]:
print("Unique categories in 'gender': ", end=" ")
print(df['gender'].unique())

print("Unique categories in 'race/ethnicity': ", end=" ")
print(df['race/ethnicity'].unique())

print("Unique categories in 'parental level of education': ", end=" ")
print(df['parental level of education'].unique())

print("Unique categories in 'lunch': ", end=" ")
print(df['lunch'].unique())

print("Unique categories in 'test preparation course': ", end=" ")
print(df['test preparation course'].unique())

Unique categories in 'gender':  ['female' 'male']
Unique categories in 'race/ethnicity':  ['group B' 'group C' 'group A' 'group D' 'group E']
Unique categories in 'parental level of education':  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Unique categories in 'lunch':  ['standard' 'free/reduced']
Unique categories in 'test preparation course':  ['none' 'completed']


**Rename column names**

In [28]:
df.rename(columns={'race/ethnicity':'race_ethnicity','parental level of education':'parental_level_of_education','test preparation course':'test_preparation_course', 'math score':'math_score','reading score':'reading_score','writing score':'writing_score'}, inplace=True)
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [29]:
df.to_csv("./data/transformed_data.csv")

**Adding Columns**

In [30]:
df['total_score'] = df['math_score'] + df['reading_score'] + df['writing_score']
df['average'] = df['total_score']/3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [31]:
reading_full = df[df['reading_score']==100]['average'].count()
writing_full = df[df['writing_score']==100]['average'].count()
math_full = df[df['math_score']==100]['average'].count()

print(f'Number of students with full marks in Maths: {math_full}')
print(f'Number of students with full marks in Reading: {reading_full}')
print(f'Number of students with full marks in Writing: {writing_full}')

Number of students with full marks in Maths: 7
Number of students with full marks in Reading: 17
Number of students with full marks in Writing: 14


In [32]:
reading_less_20 = df[df['reading_score']<=20]['average'].count()
writing_less_20 = df[df['writing_score']<=20]['average'].count()
math_less_20 = df[df['math_score']<=20]['average'].count()

print(f'Number of students with less than 20 marks in Maths: {math_less_20}')
print(f'Number of students with less than 20 marks in Reading: {reading_less_20}')
print(f'Number of students with less than 20 marks in Writing: {writing_less_20}')

Number of students with less than 20 marks in Maths: 4
Number of students with less than 20 marks in Reading: 1
Number of students with less than 20 marks in Writing: 3


**Insights**
1. More number of students have performed the worst in Maths
2. Best performance is in reading section